### PR TITLE
Exception support

### PR DESCRIPTION
**Pull request by terrafrost**
Crypt/Base.php now uses exceptions.

To accomodate this decrypt() can no longer return false. If padding is enabled and the padding length is invalid (the only condition under which false was returned) an InvalidInputException will be thrown.

Also, previously, when encrypting, if the padding was disabled and the plaintext wasn't a multiple of the block size a E_USER_NOTICE would be issued via user_error() and then padding would be enabled. Now it's not enabled - an exception is thrown suggesting that you might want to enable padding and that's that.

I figure I'll add exceptions to more classes based on the feedback I get to this commit.

This commit should not be merged into either 1.0 or 2.0 - just master and maybe then a new 2.1 branch when the time is right.

I like exceptions over user_error's because they show the stack trace. With user_error's you just see the line the error was issued on but that's not that helpful since the relevant line for the user is the line where they called the method that triggered the error - not a line in a method that was called by the method that was called by the method that the user called.

Also, in some cases, I think catching errors can be less cumbersome than trying to use an error handler. I mean, if you just want to blanket suppress all errors then adjusting the error_reporting level is even easier than throwing the whole code into a giant try / catch block but idk...  mixing user_error's and try / catch seems...  tacky. Seems like it's better to use all one or all of the other - not both.

### CODE REVIEW

**Comment by bantu** [comment]
We might have to look into (whether and) how to properly document Exceptions so they show up in the API docs.

**Comment by terrafrost** [comment]
The latest commit adds `@throw` to the docblock comment and replaces the user-defined exceptions with SPL exceptions.

**Comment by terrafrost** [comment]
In case anyone is curious Laravel throws a mix of SPL Exceptions and custom exceptions. Here's the (annotated) output of `grep -r "throw new"` (non-SPL exceptions have `**`'s prepended before them):

```
./tests/View/fixtures/section-exception.php:<?php throw new Exception('section exception message') ?>
** ./tests/Routing/RoutingRouteTest.php:        $router->get('foo/bar', function () { throw new Illuminate\Http\Exception\HttpResponseException(new Response('hello')); });
./tests/Queue/QueueWorkerTest.php:        $job->shouldReceive('fire')->once()->andReturnUsing(function () { throw new RuntimeException; });
./tests/Queue/QueueWorkerTest.php:        $job->shouldReceive('fire')->once()->andReturnUsing(function () { throw new RuntimeException; });
./tests/Queue/QueueSyncQueueTest.php:        throw new Exception();
./tests/Database/DatabaseConnectionTest.php:            $mock->transaction(function () { throw new Exception('foo'); });
./tests/Cache/CacheDatabaseStoreTest.php:            throw new Exception;
./src/Illuminate/View/View.php:        throw new BadMethodCallException("Method [$method] does not exist on view.");
./src/Illuminate/View/FileViewFinder.php:            throw new InvalidArgumentException("View [$name] has an invalid name.");
./src/Illuminate/View/FileViewFinder.php:            throw new InvalidArgumentException("No hint path defined for [{$segments[0]}].");
./src/Illuminate/View/FileViewFinder.php:        throw new InvalidArgumentException("View [$name] not found.");
./src/Illuminate/View/Factory.php:            throw new InvalidArgumentException("Unrecognized extension in file: $path");
./src/Illuminate/View/Engines/EngineResolver.php:        throw new InvalidArgumentException("Engine $engine not found.");
./src/Illuminate/Validation/Validator.php:            throw new InvalidArgumentException('Attribute for each() must be an array.');
./src/Illuminate/Validation/Validator.php:            throw new RuntimeException('Presence verifier has not been set.');
./src/Illuminate/Validation/Validator.php:            throw new InvalidArgumentException("Validation rule $rule requires at least $count parameters.");
./src/Illuminate/Validation/Validator.php:        throw new BadMethodCallException("Method [$method] does not exist.");
** ./src/Illuminate/Validation/ValidatesWhenResolvedTrait.php:        throw new ValidationException($validator);
** ./src/Illuminate/Validation/ValidatesWhenResolvedTrait.php:        throw new UnauthorizedException;
./src/Illuminate/Support/Traits/Macroable.php:        throw new BadMethodCallException("Method {$method} does not exist.");
./src/Illuminate/Support/Traits/Macroable.php:        throw new BadMethodCallException("Method {$method} does not exist.");
./src/Illuminate/Support/Str.php:                throw new RuntimeException('Unable to generate random string.');
./src/Illuminate/Support/Str.php:            throw new RuntimeException('OpenSSL extension is required for PHP 5 users.');
./src/Illuminate/Support/ServiceProvider.php:        throw new BadMethodCallException("Call to undefined method [{$method}]");
./src/Illuminate/Support/Manager.php:        throw new InvalidArgumentException("Driver [$driver] not supported.");
./src/Illuminate/Support/Facades/Facade.php:        throw new RuntimeException('Facade does not implement getFacadeAccessor method.');
./src/Illuminate/Support/Collection.php:            throw new InvalidArgumentException("You requested {$amount} items, but there are only {$count} items in the collection");
./src/Illuminate/Session/Store.php:            throw new InvalidArgumentException('Bag not registered.');
./src/Illuminate/Routing/UrlGenerator.php:        throw new InvalidArgumentException("Route [{$name}] not defined.");
./src/Illuminate/Routing/UrlGenerator.php:        throw new InvalidArgumentException("Action {$action} not defined.");
** ./src/Illuminate/Routing/Router.php:            throw new NotFoundHttpException;
** ./src/Illuminate/Routing/RouteCollection.php:        throw new NotFoundHttpException;
** ./src/Illuminate/Routing/RouteCollection.php:        throw new MethodNotAllowedHttpException($others);
** ./src/Illuminate/Routing/Route.php:            throw new NotFoundHttpException;
./src/Illuminate/Routing/Route.php:        throw new LogicException('Route is not bound.');
./src/Illuminate/Routing/Route.php:            throw new UnexpectedValueException(sprintf(
./src/Illuminate/Routing/Route.php:            throw new LogicException("Unable to prepare route [{$this->uri}] for serialization. Uses Closure.");
./src/Illuminate/Routing/Controller.php:            throw new InvalidArgumentException("Filter method [$filter] does not exist.");
** ./src/Illuminate/Routing/Controller.php:        throw new NotFoundHttpException('Controller method not found.');
./src/Illuminate/Routing/Controller.php:        throw new BadMethodCallException("Method [$method] does not exist.");
./src/Illuminate/Queue/QueueManager.php:        throw new InvalidArgumentException("No connector for [$driver]");
./src/Illuminate/Queue/Queue.php:        throw new RuntimeException('Push queues only supported by Iron.');
./src/Illuminate/Queue/Console/SubscribeCommand.php:            throw new RuntimeException('Iron.io based queue must be default.');
./src/Illuminate/Mail/Mailer.php:        throw new InvalidArgumentException('Invalid view.');
./src/Illuminate/Mail/Mailer.php:        throw new InvalidArgumentException('Callback is not valid.');
./src/Illuminate/Log/Writer.php:            throw new RuntimeException('Events dispatcher has not been set.');
./src/Illuminate/Log/Writer.php:        throw new InvalidArgumentException('Invalid log level.');
./src/Illuminate/Http/Request.php:            throw new RuntimeException('Session store not set on request.');
./src/Illuminate/Http/RedirectResponse.php:        throw new BadMethodCallException("Method [$method] does not exist on Redirect.");
./src/Illuminate/Hashing/BcryptHasher.php:            throw new RuntimeException('Bcrypt hashing not supported.');
./src/Illuminate/Foundation/helpers.php:        throw new RuntimeException('Application session store not set.');
./src/Illuminate/Foundation/helpers.php:        throw new InvalidArgumentException("File {$file} not defined in asset manifest.");
** ./src/Illuminate/Foundation/Validation/ValidatesRequests.php:        throw new HttpResponseException($this->buildFailedValidationResponse(
./src/Illuminate/Foundation/Testing/WithoutMiddleware.php:            throw new Exception('Unable to disable middleware. CrawlerTrait not used.');
./src/Illuminate/Foundation/Testing/WithoutEvents.php:            throw new Exception('Unable to disable middleware. ApplicationTrait not used.');
** ./src/Illuminate/Foundation/Testing/CrawlerTrait.php:            throw new PHPUnitException($message, null, $this->response->exception);
./src/Illuminate/Foundation/Testing/CrawlerTrait.php:                throw new InvalidArgumentException(
./src/Illuminate/Foundation/Testing/CrawlerTrait.php:            throw new InvalidArgumentException(
./src/Illuminate/Foundation/Testing/CrawlerTrait.php:            throw new InvalidArgumentException(
./src/Illuminate/Foundation/Testing/ApplicationTrait.php:                throw new Exception(
** ./src/Illuminate/Foundation/Http/Middleware/VerifyPostSize.php:            throw new PostTooLargeException;
** ./src/Illuminate/Foundation/Http/Middleware/VerifyCsrfToken.php:        throw new TokenMismatchException;
** ./src/Illuminate/Foundation/Http/Middleware/CheckForMaintenanceMode.php:            throw new HttpException(503);
** ./src/Illuminate/Foundation/Http/FormRequest.php:        throw new HttpResponseException($this->response(
** ./src/Illuminate/Foundation/Http/FormRequest.php:        throw new HttpResponseException($this->forbiddenResponse());
./src/Illuminate/Foundation/Console/ServeCommand.php:                throw new Exception("HHVM's built-in server requires HHVM >= 3.8.0.");
** ./src/Illuminate/Foundation/Bootstrap/HandleExceptions.php:            throw new ErrorException($message, 0, $level, $file, $line);
** ./src/Illuminate/Foundation/Auth/ResetsPasswords.php:            throw new NotFoundHttpException;
** ./src/Illuminate/Foundation/Application.php:            throw new NotFoundHttpException($message);
** ./src/Illuminate/Foundation/Application.php:        throw new HttpException($code, $message, null, $headers);
./src/Illuminate/Foundation/Application.php:        throw new RuntimeException('Unable to detect application namespace.');
** ./src/Illuminate/Filesystem/FilesystemAdapter.php:            throw new ContractFileNotFoundException($path, $e->getCode(), $e);
./src/Illuminate/Filesystem/FilesystemAdapter.php:        throw new InvalidArgumentException('Unknown visibility: '.$visibility);
** ./src/Illuminate/Filesystem/Filesystem.php:        throw new FileNotFoundException("File does not exist at path {$path}");
** ./src/Illuminate/Filesystem/Filesystem.php:        throw new FileNotFoundException("File does not exist at path {$path}");
** ./src/Illuminate/Encryption/McryptEncrypter.php:            throw new RuntimeException('The only supported ciphers are MCRYPT_RIJNDAEL_128 and MCRYPT_RIJNDAEL_256.');
** ./src/Illuminate/Encryption/McryptEncrypter.php:            throw new DecryptException($e->getMessage());
./src/Illuminate/Encryption/EncryptionServiceProvider.php:                throw new RuntimeException('No supported encrypter found. The cipher and / or key length are invalid.');
./src/Illuminate/Encryption/Encrypter.php:            throw new RuntimeException('The only supported ciphers are AES-128-CBC and AES-256-CBC with the correct key lengths.');
** ./src/Illuminate/Encryption/Encrypter.php:            throw new EncryptException('Could not encrypt the data.');
** ./src/Illuminate/Encryption/Encrypter.php:            throw new DecryptException('Could not decrypt the data.');
** ./src/Illuminate/Encryption/BaseEncrypter.php:            throw new DecryptException('The payload is invalid.');
** ./src/Illuminate/Encryption/BaseEncrypter.php:            throw new DecryptException('The MAC is invalid.');
./src/Illuminate/Database/Query/Builder.php:            throw new InvalidArgumentException;
./src/Illuminate/Database/Query/Builder.php:            throw new InvalidArgumentException('Illegal operator and value combination.');
./src/Illuminate/Database/Query/Builder.php:            throw new InvalidArgumentException("Invalid binding type: {$type}.");
./src/Illuminate/Database/Query/Builder.php:            throw new InvalidArgumentException("Invalid binding type: {$type}.");
./src/Illuminate/Database/Query/Builder.php:        throw new BadMethodCallException("Call to undefined method {$className}::{$method}()");
** ./src/Illuminate/Database/Eloquent/Relations/BelongsToMany.php:        throw new ModelNotFoundException;
** ./src/Illuminate/Database/Eloquent/QueueEntityResolver.php:        throw new EntityNotFoundException($type, $id);
** ./src/Illuminate/Database/Eloquent/Model.php:                throw new MassAssignmentException($key);
./src/Illuminate/Database/Eloquent/Model.php:            throw new Exception('No primary key defined on model.');
./src/Illuminate/Database/Eloquent/Model.php:            throw new LogicException('Relationship method must return an object of type '
./src/Illuminate/Database/Eloquent/FactoryBuilder.php:                throw new InvalidArgumentException("Unable to locate factory with name [{$this->name}].");
./src/Illuminate/Database/DatabaseManager.php:            throw new InvalidArgumentException("Database [$name] not configured.");
./src/Illuminate/Database/Connectors/SQLiteConnector.php:            throw new InvalidArgumentException("Database (${config['database']}) does not exist.");
./src/Illuminate/Database/Connectors/ConnectionFactory.php:            throw new InvalidArgumentException('A driver must be specified.');
./src/Illuminate/Database/Connectors/ConnectionFactory.php:        throw new InvalidArgumentException("Unsupported driver [{$config['driver']}]");
./src/Illuminate/Database/Connectors/ConnectionFactory.php:        throw new InvalidArgumentException("Unsupported driver [$driver]");
** ./src/Illuminate/Database/Connection.php:            throw new QueryException(
./src/Illuminate/Database/Connection.php:        throw new LogicException('Lost connection and no reconnector available.');
./src/Illuminate/Database/Connection.php:            throw new RuntimeException("Can't swap PDO instance while within transaction.");
./src/Illuminate/Container/Container.php:            throw new InvalidArgumentException('Method not provided.');
** ./src/Illuminate/Container/Container.php:            throw new BindingResolutionContractException($message);
** ./src/Illuminate/Container/Container.php:        throw new BindingResolutionContractException($message);
./src/Illuminate/Console/Scheduling/Event.php:            throw new LogicException('Must direct output to a file in order to e-mail results.');
./src/Illuminate/Console/Scheduling/CallbackEvent.php:            throw new InvalidArgumentException(
./src/Illuminate/Console/Scheduling/CallbackEvent.php:            throw new LogicException(
./src/Illuminate/Console/Parser.php:            throw new InvalidArgumentException('Console command definition is empty.');
./src/Illuminate/Console/Parser.php:            throw new InvalidArgumentException('Unable to determine command name from signature.');
./src/Illuminate/Cache/MemcachedConnector.php:            throw new RuntimeException('No Memcached servers added.');
./src/Illuminate/Cache/MemcachedConnector.php:            throw new RuntimeException('Could not establish Memcached connection.');
./src/Illuminate/Cache/CacheManager.php:            throw new InvalidArgumentException("Cache store [{$name}] is not defined.");
** ./src/Illuminate/Bus/MarshalException.php:        throw new static("Unable to map parameter [{$parameter->name}] to command [{$command}]");
./src/Illuminate/Bus/Dispatcher.php:            throw new RuntimeException('Queue resolver did not return a Queue implementation.');
./src/Illuminate/Bus/Dispatcher.php:        throw new InvalidArgumentException("No handler registered for command [{$className}]");
./src/Illuminate/Broadcasting/BroadcastManager.php:            throw new InvalidArgumentException("Broadcaster [{$name}] is not defined.");
./src/Illuminate/Auth/Passwords/PasswordBroker.php:            throw new UnexpectedValueException('User must implement CanResetPassword interface.');
./src/Illuminate/Auth/Guard.php:            throw new RuntimeException('Cookie jar has not been set.');
```

**Comment by terrafrost** [comment]
Here's another version of Crypt/RSA.php with exceptions:

https://github.com/phpseclib/phpseclib-php5/commit/6ff397d8562d691b5b576e227f7296ed7f6d5d21

I did this one a while ago tho. It uses custom exceptions (unlike the last Crypt/RSA.php) and replaces returning false with throwing an exception. Per http://codeutopia.net/blog/2010/03/11/should-a-failed-function-return-a-value-or-throw-an-exception/ "exceptions are for exceptional conditions". Maybe some of the exceptions that even this current incarnation throw aren't "exceptional" but [RFC3447](https://tools.ietf.org/html/rfc3447) says errors are to be output, which is exactly what's being done.

**Comment by metaclassing** [comment]
Some silly questions just popped into my head:

Given how PHP7 seems to be changing exception and general runtime error handling, does it make sense to target the exception implementation for 7's looming launch?

Would you want master's catches to all include Throwable/Error as well as Throwable/Exception day one? Or is that only going to be added later after the entire codebase gets a PHP7 scrub?

I am still a little unclear on how 7 will change implementations. When dependent components like SSH2 leverages RSA: Does SSH2 wrap calls to algorithms in try blocks and catch the RSA throwables just to turn around and throw a more specific SSH exception?

**Comment by terrafrost** [comment]
You're referring to this?: https://wiki.php.net/rfc/throwable-interface

Anyway I don't know that I see a lot of need to throw Error's vs Exception's. Seems like throwing Exception's even in an environment where Error's are now a throwable thing is perfectly fine idk. But who knows how what people consider to be PHP best practices will evolve.

> Does SSH2 wrap calls to algorithms in try blocks and catch the RSA throwables just to turn around and throw a more specific SSH exception?

Currently SSH2 does not catch and rethrow exceptions. Mostly I expect people will just be doing `try {} catch (Exception $e) {}`. There are a few exceptions I could see people /really/ wanting to catch (`\UnsupportedAlgorithmException` and `\NoSupportedAlgorithmsException`) but beyond that idk that I really see an overwhelming need to catch `\RuntimeException` vs `\Exception`.

File: phpseclib/Crypt/RSA.php [hunk 1]
```diff
@@ -2753,8 +2747,7 @@ function _rsassa_pkcs1_v1_5_verify($m, $s)
         // Length checking
 
         if (strlen($s) != $this->k) {
-            user_error('Invalid signature');
-            return false;
+            throw new \RuntimeSignature('Invalid signature');
```
Comment (bantu) [inline phpseclib/Crypt/RSA.php]: RuntimeSignature :-P - same below

Comment (terrafrost) [inline phpseclib/Crypt/RSA.php]: Good...  catch. lol

**Comment by terrafrost** [comment]
This is ready to be merged.

**Pull request merged**